pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go

```diff
@@ -0,0 +1,262 @@
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpscheme "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/scheme"
+	kcpinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	initializationbootstrap "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization"
+	"github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/apis/initialization"
+	initializationclient "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/clientset/versioned"
+	initializationinformer "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/informers/externalversions"
+)
+
+var invalidCharacters = regexp.MustCompile(`[^a-z0-9\-]`)
+
+// Bootstrap installs the apis that drive the apibinder initializer and starts the
+// controller once they're ready.
+func Bootstrap(ctx context.Context, cfg *rest.Config, installInto logicalcluster.Name) error {
+	klog.Infof("Beginning APISet Initializer bootstrapping in cluster %q", installInto)
+	gr := metav1.GroupResource{
+		Group:    initialization.GroupName,
+		Resource: "apisets",
+	}
+	crd, err := initializationbootstrap.CustomResourceDefinition(gr)
+	if err != nil {
+		return fmt.Errorf("could not determine CustomResourceDefinition: %w", err)
+	}
+
+	resourceSchema, err := apisv1alpha1.CRDToPrefixedAPIResourceSchema(crd, invalidCharacters.ReplaceAllString(version.Get().String(), "-")) // TODO: find a sensible thing to use here
+	if err != nil {
+		return fmt.Errorf("could not convert CustomResourceDefinition to APIResourceSchema: %w", err)
+	}
+
+	scopedCfg := rest.AddUserAgent(rest.CopyConfig(cfg), "kcp-apibinder-initializer-bootstrapper")
+	kcpClusterClient, err := kcpclient.NewClusterForConfig(scopedCfg)
+	if err != nil {
+		return fmt.Errorf("could not create KCP cluster client: %w", err)
+	}
+
+	kcpClient := kcpClusterClient.Cluster(installInto)
+	apisClient := kcpClient.ApisV1alpha1()
+	resourceSchema, err = apisClient.APIResourceSchemas().Create(ctx, resourceSchema, metav1.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return fmt.Errorf("could not create APIResourceSchema: %w", err)
+	}
+	klog.Infof("Created APIResourceSchema %q|%q", installInto.String(), resourceSchema.Name)
+
+	export, err := apisClient.APIExports().Create(ctx, &apisv1alpha1.APIExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: gr.String(),
+		},
+		Spec: apisv1alpha1.APIExportSpec{
+			LatestResourceSchemas: []string{resourceSchema.Name},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return fmt.Errorf("could not create APIExport: %w", err)
+	}
+	klog.Infof("Created APIExport %q|%q", installInto.String(), export.Name)
+
+	if err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
```

Comment:
Is this part of a controller reconcile? If yes, this is not a good idea as it blocks the controller. With only limited number of workers this would break the initializer controller quickly. In other words, do the backoff outside of the reconcile through the queue.

Reply:
This one is not in a controller.